### DISCUSSION TITLE
Correct of way passing  `rngs` and `state` in submodules

### DISCUSSION THREAD

**Discussion by AakashKumarNain** [google/flax#2173] [Q&A] [answered]
I am trying to build a very simple model in flax but I don't know the right way to pass `rngs` and certain state like `deterministic` from main module to submodules. Here is how I have defined my model

```python
class SampleClassifier(nn.Module):
    def setup(self):
        self.features = FeatureExtractor()
        self.classifier = Classifier()
    
    @nn.compact
    def __call__(self, x, training):
        x = self.features(x, training)
        x = self.classifier(x, training)
        return x



class FeatureExtractor(nn.Module):
    @nn.compact
    def __call__(self, x, training):
        x = nn.Conv(64, (3, 3), padding="SAME")(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        x = nn.Dropout(0.25, deterministic=not training)(x)
        
        x = nn.Conv(64, (3, 3), padding="SAME")(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        
        x = nn.max_pool(x, (2, 2), (2, 2))
        
        x = nn.Conv(128, (3, 3), padding="SAME")(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        x = nn.Dropout(0.25, deterministic=not training)(x)
        
        x = nn.Conv(128, (3, 3), padding="SAME")(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        
        x = nn.max_pool(x, (2, 2), (2, 2))
        
        x = nn.Conv(256, (3, 3), padding="SAME")(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        x = nn.Dropout(0.25, deterministic=not training)(x)
        
        x = nn.Conv(256, (3, 3), padding="SAME")(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        x = nn.max_pool(x, (2, 2), (2, 2))
        return x
    
    
class Classifier(nn.Module):
    @nn.compact
    def __call__(self, x, training):
        # Flatten 
        x = x.reshape(x.shape[0], -1)
        
        # Dense layers
        x = nn.Dense(512)(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        x = nn.Dropout(0.5, deterministic=not training)(x)
        
        x = nn.Dense(128)(x)
        x = nn.LayerNorm()(x)
        x = nn.relu(x)
        x = nn.Dropout(0.5, deterministic=not training)(x)
        
        x = nn.Dense(10)(x)
        return x
```

1. Given that both `FeatureExtractor` and `Classifier` classes have `Dropout` layers, what's the right way to pass the `rng` for these layers from the main `SampleClassifier` class?
2. What about training state for the `Dropout` class? Do we need to follow [this](https://flax.readthedocs.io/en/latest/design_notes/arguments.html#solution) for correct dropout determinism in the submodules? Or is there an easy way to do the same?

**Comment by cgarciae** [comment, marked as answer]
Hey! Currently you have to propagate `training` manually to submodules, dropout's rng however is just passed to `apply` at the top-level using the `rngs` parameter:

```python
module = SampleClassifier()

key = jax.random.PRNGKey(0)
variables = module.init(key, x, training=False)

key = jax.random.PRNGKey(42)
y = module.apply(variables, x, training=True, rngs={'dropout': key})
```

**Reply by AakashKumarNain** [reply]
Thanks @cgarciae for the quick response. Let me try that and if it works as I am expecting, I will mark your answer. Thanks again

**Reply by marcvanzee** [reply]
@AakashKumarNain did this resolve your problem? Marking this question as answered for now but please unmark if you are still blocked.

**Reply by AakashKumarNain** [reply]
Sorry I forgot to mark it as resolved. Thanks
